File: benchmarks/javascript/mandelbrot.js
// Mandelbrot Set benchmark in JavaScript

function mandelbrot(cx, cy, maxIter) {
    let zx = 0;
    let zy = 0;
    let n = 0;
    
    while (zx * zx + zy * zy <= 4 && n < maxIter) {
        const tmp = zx * zx - zy * zy + cx;
        zy = 2 * zx * zy + cy;
        zx = tmp;
        n++;
    }
    return n;
}

function linspace(start, stop, num) {
    const step = num > 1 ? (stop - start) / (num - 1) : 0;
    const result = [];
    for (let i = 0; i < num; i++) {
        result.push(start + step * i);
    }
    return result;
}

function mandelbrotSet(xmin, xmax, ymin, ymax, width, height, maxIter) {
    const r1 = linspace(xmin, xmax, width);
    const r2 = linspace(ymin, ymax, height);
    let count = 0;
    
    for (let i = 0; i < width; i++) {
        for (let j = 0; j < height; j++) {
            const c = mandelbrot(r1[i], r2[j], maxIter);
            if (c === maxIter) {
                count++;
            }
        }
    }
    return count;
}

// Calculate mandelbrot set
const width = 200;
const height = 200;
const xmin = -2.0;
const xmax = 1.0;
const ymin = -1.5;
const ymax = 1.5;
const maxIter = 100;

const count = mandelbrotSet(xmin, xmax, ymin, ymax, width, height, maxIter);
console.log(`Mandelbrot points: ${count}`);

File: benchmarks/python/mandelbrot.py
#!/usr/bin/env python3

# Mandelbrot Set benchmark in Python

def mandelbrot(c, max_iter):
    z = 0
    n = 0
    while abs(z) <= 2 and n < max_iter:
        z = z * z + c
        n += 1
    return n

def mandelbrot_set(xmin, xmax, ymin, ymax, width, height, max_iter):
    r1 = linspace(xmin, xmax, width)
    r2 = linspace(ymin, ymax, height)
    count = 0
    for r in r1:
        for i in r2:
            c = complex(r, i)
            z = mandelbrot(c, max_iter)
            if z == max_iter:
                count += 1
    return count

def linspace(start, stop, num):
    step = (stop - start) / (num - 1) if num > 1 else 0
    return [start + step * i for i in range(num)]

# Calculate mandelbrot set
width, height = 200, 200
xmin, xmax = -2.0, 1.0
ymin, ymax = -1.5, 1.5
max_iter = 100

count = mandelbrot_set(xmin, xmax, ymin, ymax, width, height, max_iter)
print(f"Mandelbrot points: {count}")
